### PR TITLE
Show correct error when transaction failed due existance of posted one

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/35734

Resolves 

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
1) With Zcash post 2 transaction from one account. Second one should fail with a message saying only one is allowed
image
2) Transactions posted at one time but from different accounts should be successful.
3) For other coins check that "view error" button is shown. Could be tested when network is disabled before approve for example. Description should not be changed too.
4) Check that success transactions show correct message and don't show error button.

### CODE REVIEW

File: components/brave_wallet_ui/common/slices/endpoints/transaction.endpoints.ts [hunk 1]
```diff
@@ -778,7 +778,11 @@ export const transactionEndpoints = ({
     }),
 
     approveTransaction: mutation<
-      { success: boolean },
+      {
+        status: boolean
```
Comment (josheleonard) [inline components/brave_wallet_ui/common/slices/endpoints/transaction.endpoints.ts:782]: can we rename this to `success`?
```suggestion
        success: boolean
```

File: components/brave_wallet_ui/common/hooks/use-pending-transaction.ts [hunk 2]
```diff
@@ -455,23 +455,35 @@ export const usePendingTransactions = () => {
     }
 
     try {
-      await dispatch(
+      const result = await dispatch(
```
Comment (josheleonard) [inline components/brave_wallet_ui/common/hooks/use-pending-transaction.ts:458]: mind refactoring this to use `useApproveTransactionMutation` hook instead?

File: components/brave_wallet_ui/common/hooks/use-pending-transaction.ts [hunk 3]
```diff
@@ -455,23 +455,35 @@ export const usePendingTransactions = () => {
...
+              message: result.errorMessage
+            },
+            transactionId: transactionInfo.id
+          })
+        )
+      }
     } catch (error) {
       dispatch(
         UIActions.setTransactionProviderError({
           providerError: error.toString(),
```
Comment (josheleonard) [inline components/brave_wallet_ui/common/hooks/use-pending-transaction.ts:480]: should `providerError`'s value be an object now?

**Review by mkarolin — APPROVED** [review]
`strings`++

File: components/brave_wallet_ui/components/extension/post-confirmation/index.tsx [hunk 4]
```diff
@@ -170,8 +170,19 @@ export function TransactionStatus(props: Props) {
...
-    const errorDetailContent =
-      providerError && `${providerError.code}: ${providerError.message}`
+    const errorCode =
+        providerError?.code?.providerError ??
+        providerError?.code?.zcashProviderError ??
+        providerError?.code?.bitcoinProviderError ??
+        providerError?.code?.solanaProviderError ??
+        providerError?.code?.filecoinProviderError
+
+    const errorDetailContent = providerError && `${errorCode}: ${providerError.message}`
```
Comment (josheleonard) [inline components/brave_wallet_ui/components/extension/post-confirmation/index.tsx:180]: How do we want to format this string if the providerError code is not defined?

File: components/brave_wallet_ui/common/hooks/use-pending-transaction.ts [hunk 5]
```diff
@@ -455,23 +457,34 @@ export const usePendingTransactions = () => {
...
+          })
+        )
+      }
     } catch (error) {
       dispatch(
         UIActions.setTransactionProviderError({
-          providerError: error.toString(),
+          providerError: {
+            message: error.toString()
+          },
```
Comment (josheleonard) [inline components/brave_wallet_ui/common/hooks/use-pending-transaction.ts:481]: We should set the error code.

File: components/brave_wallet_ui/common/hooks/use-pending-transaction.ts [hunk 6]
```diff
@@ -455,23 +457,39 @@ export const usePendingTransactions = () => {
...
+        )
+      }
     } catch (error) {
       dispatch(
         UIActions.setTransactionProviderError({
-          providerError: error.toString(),
+          providerError: {
+            code: {
+              ...
+              emptyProviderErrorCodeUnion,
```
Comment (josheleonard) [inline components/brave_wallet_ui/common/hooks/use-pending-transaction.ts:482]: nit: spread operator should be on the same line as the object that is getting spreaded
```suggestion
              ...emptyProviderErrorCodeUnion,
```

File: components/brave_wallet_ui/components/extension/post-confirmation/index.tsx [hunk 7]
```diff
@@ -170,8 +170,20 @@ export function TransactionStatus(props: Props) {
...
+    const errorCode =
+        providerError?.code.providerError ??
+        providerError?.code.zcashProviderError ??
+        providerError?.code.bitcoinProviderError ??
+        providerError?.code.solanaProviderError ??
+        providerError?.code.filecoinProviderError
+
     const errorDetailContent =
-      providerError && `${providerError.code}: ${providerError.message}`
+        providerError && `${errorCode}: ${providerError.message}`
```
Comment (josheleonard) [inline components/brave_wallet_ui/components/extension/post-confirmation/index.tsx:181]:
```suggestion
        providerError && `${errorCode ?? BraveWallet.ProviderError.kUnknown}: ${providerError.message}`
```

Comment (cypt4) [reply components/brave_wallet_ui/components/extension/post-confirmation/index.tsx]: I moved BraveWallet.ProviderError.kUnknown to the errorCode initialization

**Review by josheleonard — APPROVED** [review]

**Review by yrliou — APPROVED** [review]
core++, I'm okay with this short-term fix, in the long-term, maybe we could improve this and users won't need to go through all UI steps again to retry the transaction.

**Review by StephenHeaps — APPROVED** [review]
iOS ++

**Pull request merged**